Typed logging with slf4j

franciscolopezsancho opened the pull request:
<!--
# Pull Request Checklist

* [ ] Have you read through the [contributor guidelines](./CONTRIBUTING.md)?
* [ ] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [ ] Have you updated the documentation?
* [ ] Have you added tests for any changed functionality?
-->
## Purpose

Change Akka ad-hoc Logger to org.slf4j.Logger

This is a gist, not a complete change. Far from it, this is just a first step towards it.

## References

https://github.com/akka/akka/issues/26537

<!--
  Are there any relevant issues / PRs / mailing lists discussions?
  Please DON`T use `Fixes` notation.
-->


## Changes

Apart from Akka ad-hoc Logger to org.slf4j.Logger important changes, I think, are worth to mention:
Stubs for Mocking the logging events also changed in StubbedActorContext. Now they are created from org.slf4j.helpers.SubstituteLoggerFactory. Not sure this can be the case. But, again, this is just an idea.
Also changed previous MDC from WithMdcBehaviorInterceptor with types [String, Any] to [String, String] Assuming this would be acceptable.


by now:
 Tests: succeeded 679, failed 92, canceled 0, ignored 0, pending 3




## Background Context

Team meeting in June, in London, where it was discussed changing current logging in typed actors

franciscolopezsancho committed:
changed Logger to slf4j in all classes but test ones. Please be aware I changed Map[String,Any] to Map[String,String] without being able to test consecuences

franciscolopezsancho committed:
added slf4j Logger on tests, no expectation by now they'll pass. Now compilation errors that comply to  convert Logger.info (debug,trace...) signature grouping params in an scala.Array.

franciscolopezsancho committed:
test running. Failing 89 of 753

franciscolopezsancho committed:
formatting with sbt

franciscolopezsancho committed:
after merge upstream

franciscolopezsancho committed:
after merge upstream. Forgot this two

johanandren commented:
OK TO TEST

johanandren commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorLoggingSpec.scala:
```diff
@@ -183,98 +172,56 @@ class ActorLoggingSpec extends ScalaTestWithActorTestKit("""
-            context.log.debug("{} {}", "arg1", "arg2")
-            context.log.debug("{} {} {}", "arg1", "arg2", "arg3")
-            context.log.debug("{} {} {} {}", "arg1", "arg2", "arg3", "arg4")
+            context.log.debug("{} {}", Array("arg1", "arg2"))
```
If I look at the Slf4J logger it seems to have overloads for two placeholder values and then one with varargs so these shouldn't be needed?

franciscolopezsancho replied:
yes, I'll fix it.

johanandren commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorLoggingSpec.scala:
```diff
@@ -1,21 +1,20 @@
 /*
- * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ * Copyright (C,cause) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
```
Ups

franciscolopezsancho replied:
Ups :|

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala:
```diff
@@ -326,13 +326,12 @@ object Behaviors {
   def withMdc[T](
       interceptMessageClass: Class[T],
-      mdcForMessage: akka.japi.function.Function[T, java.util.Map[String, Any]],
+      mdcForMessage: akka.japi.function.Function[T, java.util.Map[String, String]],
```
How about we keep this a `Map[String, Object]` and make String of the objects in the impl?

patriknw replied:
Wouldn't that be surprising for users that we are silently discarding informat (and relying on `toString`).
If it's limited to strings I think we should make that explicit all the way.

johanandren replied:
Yes, you are right, I change my mind!

franciscolopezsancho replied:
I'll leave to `Map[String, String]` then

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/ExtensionsImpl.scala:
```diff
@@ -49,7 +49,9 @@ trait ExtensionsImpl extends Extensions { self: ActorSystem[_] =>
-              log.error(problem, "While trying to load extension [{}], skipping...", extensionIdFQCN)
+              log.error(
+                "Has been problem with {}, While trying to load extension [{}], skipping...",
+                Array(problem, extensionIdFQCN))
```
If we should include `problem` it should be as the log.error exception rather than in the log string so that a stacktrace gets logged etc.

franciscolopezsancho replied:
you're right

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala:
```diff
@@ -154,15 +154,31 @@ private[akka] final class LogMessagesInterceptor(val opts: LogOptions) extends B
+        case Logging.ErrorLevel   => log.error("received message {}", msg)
+        case Logging.WarningLevel => log.warn("received message {}", msg)
+        case Logging.InfoLevel    => log.info("received message {}", msg)
+        case Logging.DebugLevel   => log.debug("received message {}", msg)
```
The type of  `opts.level` should be the native Slf4J `org.slf4j.event.Level` instead now.

patriknw replied:
That's a shortcoming to not be able to log at a given level with a parameter. Probably most annoying for us as a library and not so much for end users.

franciscolopezsancho replied:
you right `org.slf4j.event.Level` is the proper one to use.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:
```diff
@@ -81,10 +72,15 @@ private abstract class AbstractSupervisor[I, Thr <: Throwable](strategy: Supervi
+        case Logging.DebugLevel =>
+          ctx.asScala.log.debug("Supervisor {} saw failure: {}", Array(this, unwrapped.getMessage))
+        //TODO check this debug case is actually best option when other level is found
+        case _ => ctx.asScala.log.debug("Supervisor {} saw failure: {}", Array(this, unwrapped.getMessage))
```
Same here, use `org.slf4j.event.Level` and there will be no unknowns because it is an enum

franciscolopezsancho replied:
yep

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala:
```diff
@@ -120,11 +119,8 @@ private[akka] object ActorContextAdapter {
-    val logger =
-      new LoggerAdapterImpl(system.eventStream, logClass, logSource, LoggingFilterWithMarker.wrap(system.logFilter))
+  private def initLoggerWithClass(logClass: Class[_]): Logger = {
+    val logger = LoggerFactory.getLogger(logClass)
```
We also want to associate the logger with the actor path some how, that's what the `logSource` previously did.

franciscolopezsancho replied:
How you find concatenation with '/' or some other separator?

raboof replied:
It would be great if it would be possible to have it as a separate field, so when published to log infrastructure that supports structured logs (like Kibana) it is easy to group/filter/report on it

johanandren replied:
I wonder if MDC is the only option, that kind of sucks because it would mean having to always setmdc/touch thread local when actor starts running and clear after.

patriknw replied:
I agree that it should be a separate field and not part of the Logger name. The Logger name should be plain class names, since that is the normal convention with slf4j.
Not sure what possibilities there are for adding "fields"? Is that with MDC, Markers or something else?
Maybe  @wsargent can give us a hint?

Markers: https://tersesystems.com/blog/2019/05/18/application-logging-in-java-part-4/

wsargent replied:
@patriknw There are two ways you can do this using markers.  You'll always end up doing something like:

```scala
val customPathMarker: PathMarker = PathMarker(actor.path)
logger.info(customPathMarker, message)
```

but you can do that either directly, or by specifying a proxy logger that will always add the context information under the hood.

```scala
val proxyLogger: org.slf4j.Logger = PathLoggerFactory(loggerFactory, actor.path).getLogger(getClass())
proxyLogger.info(message) // will automatically add a path marker under the hood.
```

I'm still writing out the fine details, but check out 

https://github.com/tersesystems/terse-logging-apis/blob/master/logging-example-context/src/main/java/com/tersesystems/logging/example/context/AppLogger.java

and

https://github.com/tersesystems/terse-logging-apis/blob/master/logging-context-logstash/src/main/java/com/tersesystems/logging/context/logstash/AbstractLogstashLogger.java

the most general case is

https://github.com/tersesystems/terse-logging-apis/blob/master/logging-api-proxy/src/main/java/com/tersesystems/logging/api/proxy/ProxyLogger.java

I generally create these with Javapoet as it's a pretty large API:

https://github.com/tersesystems/terse-logging-apis/tree/master/logging-api-generator

patriknw replied:
thanks wsargent, we will look into that. Very useful if we can include some fields in all logging (from Loggers created by Akka)

wsargent replied:
If you're specifically interested in the actor path, you may want to create the logger based on the actorsystem + path:

```scala
val convertedPath = convertToDotFormat(actorSystem + actor.path)
val logger loggerFactory.getLogger(convertedPath)
```

You're not limited to using the class name as the logger name, it's just what most people default to.

johanandren replied:
I think the problem is that we want both

patriknw replied:
I think most users want the Logger name to correspond to package/class name (because that is what they are used to). Actor path is a good additional field but shouldn't be for the logger name.

franciscolopezsancho replied:
Outstanding info guys, thanks! 
From what I understand from above looks to me Marker is the cleaner solution? Shall I go for it?

I've been looking around regards when to use MDC vs Marker and I got from it that MDC is better suited to slice and dice or to aggregate info while Markers are more focused for filtering info, even before gets written into the log or to trigger events when some mark is found. 
Some of this info I got it from https://stackoverflow.com/questions/4165558/best-practices-for-using-markers-in-slf4j-logback which actually I found is also referred  in https://github.com/akka/akka/pull/27324#discussion_r302697422

My only concern regards using MDC, in this case, would be it looks a bit more cumbersome. But I'm far from an expert in logging. I will be fine going, either way by now, MDC or Marker, as I don't have a strong opinion about it. Please let me know what you find is best.

johanandren replied:
From my limited knowledge MDC sounds more right as it is context rather than something special with an event (like how we use markers for "Security" previously for example).

franciscolopezsancho replied:
I'm inclined to MDC as well.

franciscolopezsancho replied:
@johanandren @patriknw 
I'm a bit confused about this part, to be honest.

Let's say I add the actorPath when init the logger in this very same class: ``` 
private def initLoggerWithClass(logClass: Class[_]): Logger = {
    val logger = LoggerFactory.getLogger(logClass)
    org.slf4j.MDC.put("actorPath",self.path.name)
    actorLogger = OptionVal.Some(logger)
    logger
  }
```

What I don't quite get is at what point will I clean it. 
If (big if) I do it on akka.actor.typed.internal.adapter.ActorAdapter in .... preStart? postStop?
looks I'm splitting functionality that should be in the same place, related in a way that is easy to follow. On the contrary, this will make the code difficult to understand and maintain. 
Not only that but I'm not even sure that would fulfill the requirements. I will have to write down a test for this as I understand there's nothing currently that is testing this. 

Any suggestions are very much welcome.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorSystemAdapter.scala:
```diff
@@ -89,11 +87,7 @@ import akka.{ actor => untyped }
-    classOf[ActorSystem[_]],
-    name,
-    LoggingFilterWithMarker.wrap(untypedSystem.logFilter))
+  override val log: Logger = LoggerFactory.getLogger(classOf[ActorSystem[_]])
```
We'd also want to associate this logger with the actor system name

johanandren reviewed (commented):
Looks like a good start!

patriknw commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorLoggingSpec.scala:
```diff
@@ -183,98 +172,56 @@ class ActorLoggingSpec extends ScalaTestWithActorTestKit("""
             context.log.error(marker, "{} {} {} {} {}", Array("arg1", "arg2", "arg3", "arg4", "arg5"))
 
-            context.log.error(cause, "message")
-            context.log.error(cause, "{}", "arg1")
```
what about these with `cause`?

franciscolopezsancho replied:
I'll put it back

patriknw commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorLoggingSpec.scala:
```diff
@@ -183,98 +172,56 @@ class ActorLoggingSpec extends ScalaTestWithActorTestKit("""
-            context.log.warning(marker, "{} {} {} {} {}", Array("arg1", "arg2", "arg3", "arg4", "arg5"))
-
-            context.log.warning(cause, "message")
-            context.log.warning(cause, "{}", "arg1")
```
what about these with cause?

franciscolopezsancho replied:
I'll put it back

patriknw commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorLoggingSpec.scala:
```diff
@@ -291,12 +238,13 @@ class ActorLoggingSpec extends ScalaTestWithActorTestKit("""
-        Map("static" -> 1),
+        Map("static" -> "1"),
         // FIXME why u no infer the type here Scala??
+        //TODO review that change from Map[String,Any] to Map[String,String] is viable
```
I see, so it only has support for strings? might be a good limitation to avoid putting mutable things in there

franciscolopezsancho replied:
Only strings, yes. I agree. I think it would be more robust and also simpler (at least to maintain)

patriknw commented on akka-actor-typed-tests/src/test/scala/docs/akka/typed/SpawnProtocolDocSpec.scala:
```diff
@@ -64,7 +64,7 @@ class SpawnProtocolDocSpec extends ScalaTestWithActorTestKit with WordSpecLike {
 
       val greetedBehavior = Behaviors.receive[HelloWorld.Greeted] { (context, message) =>
-        context.log.info("Greeting for {} from {}", message.whom, message.from)
+        context.log.info("Greeting for {} from {}", Array(message.whom, message.from))
```
didn't it work with plain parameters (varargs)?

franciscolopezsancho replied:
I got this error:

```ambiguous reference to overloaded definition,
[error] both method info in trait Logger of type (x$1: String, x$2: Object*)Unit
[error] and  method info in trait Logger of type (x$1: String, x$2: Any, x$3: Any)Unit```

so now I changed it to:
``` context.log.info("Greeting for {} from {}", Array(message.whom, message.from):_*)```


sorry 
``` context.log.info("Greeting for {} from {}", Array(message.whom, message.from))``` was wrong

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:
```diff
@@ -81,10 +72,15 @@ private abstract class AbstractSupervisor[I, Thr <: Throwable](strategy: Supervi
       strategy.logLevel match {
         case Logging.ErrorLevel =>
-          ctx.asScala.log.error(unwrapped, "Supervisor {} saw failure: {}", this, unwrapped.getMessage)
+          ctx.asScala.log.error("Supervisor {} saw failure: {}", Array(this, unwrapped.getMessage))
```
For those we need to include the exception, using:
```
public void error(String msg, Throwable t)
```

which doesn't have support for variable substitutions, but since it's on error level it's fine to eagerly replace them with Scala string interpolation.

```
ctx.asScala.log.error(s"Supervisor $this saw failure: ${unwrapped.getMessage}", unwrapped)
```

franciscolopezsancho replied:
you're right

franciscolopezsancho replied:
I think I will go for @patriknw :
```ctx.asScala.log.error(s"Supervisor $this saw failure:", unwrapped)```

patriknw reviewed (commented):
@franciscolopezsancho Excellent that you have started this.

franciscolopezsancho committed:
changed after comments in https://github.com/akka/akka/pull/27324. All comments addresed but MDC to log actorpath

franciscolopezsancho committed:
Merge branch 'master' of https://github.com/akka/akka into typed-logging-with-slf4j

franciscolopezsancho committed:
formatting

franciscolopezsancho committed:
added logback as default log4j in tests, plus some minor changes to allow compilation. Also added to test MDC put

franciscolopezsancho committed:
Merge branch 'master' of https://github.com/akka/akka into typed-logging-with-slf4j

johanandren commented on akka-actor-typed-tests/src/test/scala/docs/akka/typed/IntroSpec.scala:
```diff
@@ -186,7 +186,7 @@ object IntroSpec {
             Behaviors.same
           case MessagePosted(screenName, message) =>
-            context.log.info("message has been posted by '{}': {}", Array(screenName, message))
+            context.log.info("message has been posted by '{}': {}", Array(screenName, message): _*)
```
Why create the array at all, doesn't this work:
`context.log.info("message has been posted by '{}': {}", screenName, message)`

franciscolopezsancho replied:
This is the only way I found out to tell the compiler we had a specific signature. Maybe there's something better. That I don't know by now.

johanandren replied:
Oh, that's pretty bad that we get that from Scala.

wsargent commented:
You get that in Java as well, any time you pass in a pre created array instead of varargs.  You have to cast using (Object[]).

> On Jul 16, 2019, at 2:22 AM, Johan Andrén <notifications@github.com> wrote:
> 
> @johanandren commented on this pull request.
> 
> In akka-actor-typed-tests/src/test/scala/docs/akka/typed/IntroSpec.scala:
> 
> > @@ -186,7 +186,7 @@ object IntroSpec {
>              handle ! PostMessage("Hello World!")
>              Behaviors.same
>            case MessagePosted(screenName, message) =>
> -            context.log.info("message has been posted by '{}': {}", Array(screenName, message))
> +            context.log.info("message has been posted by '{}': {}", Array(screenName, message): _*)
> Oh, that's pretty bad that we get that from Scala.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

raboof commented:
For "context.log.info("message has been posted by '{}': {}",
Array(screenName, message))" it makes sense, but for "context.log.info("message
has been posted by '{}': {}", screenName, message)" it surprised me. Do you
have a nice way to avoid that?

On Tue, Jul 16, 2019 at 5:18 PM Will Sargent <notifications@github.com>
wrote:

> You get that in Java as well, any time you pass in a pre created array
> instead of varargs. You have to cast using (Object[]).
>
> > On Jul 16, 2019, at 2:22 AM, Johan Andrén <notifications@github.com>
> wrote:
> >
> > @johanandren commented on this pull request.
> >
> > In akka-actor-typed-tests/src/test/scala/docs/akka/typed/IntroSpec.scala:
> >
> > > @@ -186,7 +186,7 @@ object IntroSpec {
> > handle ! PostMessage("Hello World!")
> > Behaviors.same
> > case MessagePosted(screenName, message) =>
> > - context.log.info("message has been posted by '{}': {}",
> Array(screenName, message))
> > + context.log.info("message has been posted by '{}': {}",
> Array(screenName, message): _*)
> > Oh, that's pretty bad that we get that from Scala.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka/pull/27324?email_source=notifications&email_token=AABAGEHGUWPS3RUL4CY5AODP7XRD7A5CNFSM4H7NZZIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2BGJZI#issuecomment-511861989>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABAGED77AJEON4XRX22PB3P7XRD7ANCNFSM4H7NZZIA>
> .
>

wsargent commented:
```scala
context.log.info(s"message has been posted by '${screenName}': ${message}")
```

:-)

franciscolopezsancho committed:
adding appender from logback to intercept events. This is just to check the gist of using logback's appender to capture events

franciscolopezsancho committed:
Merge branch 'master' of https://github.com/akka/akka into typed-logging-with-slf4j

franciscolopezsancho commented:
Good morning. I've been hacking to intercept events and pass at least one test of `akka.actor.typed.LogMessagesSpec` 
I'd like to verify that my assumptions are agreed before going any further. 
1. In order to intercept an event, I did use Logback class ILoggingEvent instead of SLF4J class LoggingEvent. This is my main concern. 
2. SLF4J couldn't hold the source of the call to the log, this means the actor from where logging happens. To overcome this I added this reference to the actor in the message log itself. 
3. I added the appender at running time by now but not planning to leave it like this. I'll move it to an XML when cleaning.

wsargent commented:
> In order to intercept an event, I did use Logback class ILoggingEvent instead of SLF4J class LoggingEvent. This is my main concern.

I don't see a problem with that as long as it's in the specs, since you need to have some kind of SLF4J implementation to be useful.  If it were in testkit, then it would be a problem.

> SLF4J couldn't hold the source of the call to the log, this means the actor from where logging happens. To overcome this I added this reference to the actor in the message log itself.

Can you point to a line instance?  

Also, I've been trying to come up with a better definition of "debug logging for an audience" and so far am describing it as "diagnostic logging" -- more in [blog post here](https://tersesystems.com/blog/2019/07/22/targeted-diagnostic-logging-in-production/).

franciscolopezsancho committed:
formatting

franciscolopezsancho commented:
@wsargent Thanks for having a look into this. 
1. I did use Logback in here 
https://github.com/franciscolopezsancho/akka/blob/b45a7827bc4fdc8da0c40f99295f401bd867c4da/akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala#L77

which in a way is like testkit. This class is a new implementation of https://github.com/franciscolopezsancho/akka/blob/typed-logging-with-slf4j/akka-testkit/src/main/scala/akka/testkit/TestEventListener.scala because this class already is dependent of `akka.event.logging.LogEvent` as per https://github.com/franciscolopezsancho/akka/blob/b45a7827bc4fdc8da0c40f99295f401bd867c4da/akka-testkit/src/main/scala/akka/testkit/TestEventListener.scala#L87


Also, I included the appender as parameter https://github.com/franciscolopezsancho/akka/blob/b45a7827bc4fdc8da0c40f99295f401bd867c4da/akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala#L54 when intercepting but that may be avoidable. I just thought that keeping it in the signature would be more explicit. 

2. I did add the actor address in here https://github.com/franciscolopezsancho/akka/blob/b45a7827bc4fdc8da0c40f99295f401bd867c4da/akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala#L178  as this is just used in typed actors. This way will be able to check the actor's path. 

Very interesting article, It is a great idea. Log at debug level while keeping filters in place to avoid that much overload (TurboFilter will do even before the log event is processed, so will be performant). When you need to debug the system in prod you just need to tune the filters while looking for the root of the problem. Did I got it right?

wsargent commented:
Ah, okay -- I think you should be using `org.slf4j.event.LoggingEvent` in anything that's a public API.  Avoid `ListAppender` as it's logback-specific -- I think you can use `SubstituteLogger` and a logback delegate in tests if that's what you're using it for.

> Log at debug level while keeping filters in place to avoid that much overload 

You can log at info level, because turbofilter gets run for all events, debug or not: but the only thing it does is delegate processing to the marker passed in with the log statement.  The marker can have access to specific user/session data and can factor that into the decision to log, and can also call out to actors / JMX / feature flags, etc.

franciscolopezsancho commented:
Great! Thanks, @wsargent. This is really helpful.

franciscolopezsancho committed:
changed specific logback Logger and Event for SLF4J. All test from LogMessagesSpec passing which include MDC

franciscolopezsancho committed:
Merge branch 'master' of https://github.com/akka/akka into typed-logging-with-slf4j

franciscolopezsancho commented:
Hi, @wsargent. I already changed all logback specific to SLF4J following your advice and tips. Many thanks 👍 With that I managed to pass all tests from `akka.actor.typed.LogMessagesSpec` that includes MDC. 

Could anybody have a look at it, please? I can jump into a call if want to, that may be more convenient for any of you. I added a brief description below:

 I changed EventFilter.intercept signature. Actually, I created one specific to typed actor (`akka.actor.typed.testkit.EventFilter`)  as the already existing `akka.testkit.EventFilter` is used in many places not typed actor related. 

I just have a few questions before I can move forward and expand what I used in `akka.actor.typed.LogMessagesSpec` to the rest of typed-actor tests. 

1.  Is this signature https://github.com/franciscolopezsancho/akka/blob/03c89ec72327f6f097f32d83eff4b129473d93dd/akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala#L54 good enough? Previously event where registered, I think, using publish/subscribe on  `system.eventStream` in line https://github.com/franciscolopezsancho/akka/blob/03c89ec72327f6f097f32d83eff4b129473d93dd/akka-testkit/src/main/scala/akka/testkit/TestEventListener.scala#L116

2. Do you find appropriate the testing of MDC I used in `akka.actor.typed.LogMessagesSpec`

3. Is there anything else you find I'm missing based on passing the tests on https://github.com/franciscolopezsancho/akka/blob/typed-logging-with-slf4j/akka-actor-typed-tests/src/test/scala/akka/actor/typed/LogMessagesSpec.scala?


Thanks for all your support

wsargent commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala:
```diff
@@ -0,0 +1,474 @@
+   * Apply this filter while executing the given code block. Care is taken to
+   * remove the filter when the block is finished or aborted.
+   */
+  def intercept[T](code: => T, eventsQueue: LinkedBlockingQueue[SubstituteLoggingEvent])(
```
```suggestion
  def intercept[T, LE <: LoggingEvent](code: => T, eventsQueue: BlockingQueue[LE])(
```

wsargent commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala:
```diff
@@ -0,0 +1,474 @@
+  def intercept[T](code: => T, eventsQueue: LinkedBlockingQueue[SubstituteLoggingEvent])(
+      implicit system: ActorSystem): T = {
+    //TODO add appender and clean it @see original TestEventListener
+    def leftToDo: Int = todo - eventsQueue.asScala.filter(each => matches(each)).size
```
```suggestion
        def leftToDo: Int = todo - eventsQueue.asScala.count(matches)
```

wsargent reviewed (commented):
I've pulled the code and am starting a review, but I'm probably not going to finish it tonight.  Will poke at it some more.

wsargent commented:
Got a little further today, will see about reviewing MDC stuff tonight

On Mon, Jul 29, 2019 at 5:45 AM franciscolopezsancho <
notifications@github.com> wrote:

> Hi, @wsargent <https://github.com/wsargent>. I already changed all
> logback specific to SLF4J following your advice and tips. Many thanks 👍
> With that I managed to pass all tests from
> akka.actor.typed.LogMessagesSpec that includes MDC.
>
> Could anybody have a look at it, please? I can jump into a call if want
> to, that may be more convenient for any of you. I added a brief description
> below:
>
> I changed EventFilter.intercept signature. Actually, I created one
> specific to typed actor (akka.actor.typed.testkit.EventFilter) as the
> already existing akka.testkit.EventFilter is used in many places not
> typed actor related.
>
> I just have a few questions before I can move forward and expand what I
> used in akka.actor.typed.LogMessagesSpec to the rest of typed-actor tests.
>
>    1.
>
>    Is this signature
>    https://github.com/franciscolopezsancho/akka/blob/03c89ec72327f6f097f32d83eff4b129473d93dd/akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala#L54
>    good enough? Previously event where registered, I think, using
>    publish/subscribe on system.eventStream in line
>    https://github.com/franciscolopezsancho/akka/blob/03c89ec72327f6f097f32d83eff4b129473d93dd/akka-testkit/src/main/scala/akka/testkit/TestEventListener.scala#L116
>    2.
>
>    Do you find appropriate the testing of MDC I used in
>    akka.actor.typed.LogMessagesSpec
>    3.
>
>    Is there anything else you find I'm missing based on passing the tests
>    on
>    https://github.com/franciscolopezsancho/akka/blob/typed-logging-with-slf4j/akka-actor-typed-tests/src/test/scala/akka/actor/typed/LogMessagesSpec.scala
>    ?
>
> Thanks for all your support
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka/pull/27324?email_source=notifications&email_token=AAARMRG6BYURCG3OEBKIVVDQB3Q7FA5CNFSM4H7NZZIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3ASVMA#issuecomment-515975856>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAARMRC37RZ2OYRLDD7T2ALQB3Q7FANCNFSM4H7NZZIA>
> .
>

franciscolopezsancho commented:
:) Thanks for having a look at it, @wsargent

wsargent reviewed (commented):
Looks good, I would add tests to ensure that you can have two actors with different MDC and the MDC data doesn't cross over the async boundary

franciscolopezsancho commented:
This is great, @wsargent . I'll add these suggestions. Many thanks!!

franciscolopezsancho commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala:
```diff
@@ -0,0 +1,474 @@
+  def matches(event: LoggingEvent) = {
+    event.getLevel match {
+      case Level.DEBUG => doMatch(event.getMessage)
+      case Level.TRACE => doMatch(event.getMessage) //this is a bug in slf4j, already requested a PR to fix it. https://github.com/qos-ch/slf4j/pull/211
```
: ))

franciscolopezsancho replied:
Fixed in 1.7.27

franciscolopezsancho commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala:
```diff
@@ -171,6 +195,12 @@ private[akka] final class LogMessagesInterceptor(val opts: LogOptions) extends B
+
+  def addMDC(msg: String): String = {
+    if (MDC.getMDCAdapter.getCopyOfContextMap != null)
+      msg + s" MDC is ${MDC.getMDCAdapter.getCopyOfContextMap.asScala}"
```
I already fixed most of the issues regards the last group of suggestions, but this one is hard as it is entangled with the config that is passed to the `ScalaTestWithActorTestKit`. That is `akka.testkit.TestEventListener`. If I change `akka.testkit.TestEventListener` to `akka.event.slf4j.Slf4jLogger` then I can use MDC from logback-test.xml but then `loggingEvent` is not intercepted.  I'll keep working a bit more on it, and if I can't solve it, I'll push all the changes regardless.

franciscolopezsancho replied:
Forget about it. I think I found the solution :)

franciscolopezsancho replied:
I was wrong, my bad. I did have to stick to add MDC on the `InterceptorImpl, after all. Couldn't apply an appender, to the `SubstituteLogger` that would insert the MDC as you suggested. Example of this is when I'm passing the `SubstituteLogger` as LogOptions on test `log messages with provided log level` on class `LogMessagesSpec`.

I'll keep trying as long as I keep going forward, making most of the test pass. There's still quite a bit to get things done
`[info] Tests: succeeded 681, failed 87, canceled 0, ignored 0, pending 15` on Akka-actor-typed-tests

wsargent commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala:
```diff
@@ -0,0 +1,474 @@
+   * `null` does NOT work (passing `null` disables the
+   * source filter).''
+   */
+  def apply[A <: Throwable: ClassTag](
```
Also I suppose technically you can say filter based on an argument -- or anything from `ILoggingEvent`

franciscolopezsancho replied:
I added a reference to `org.slf4j.event.LoggingEvent` as suggested 

would that suffice, you think?

Also `(implicit ev: A = Logging.Error.NoCause.getClass)`

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:
```diff
@@ -80,11 +72,16 @@ private abstract class AbstractSupervisor[I, Thr <: Throwable](strategy: Supervi
-          ctx.asScala.log.warning(unwrapped, "Supervisor {} saw failure: {}", this, unwrapped.getMessage)
-        case level => ctx.asScala.log.log(level, "Supervisor {} saw failure: {}", this, unwrapped.getMessage)
+        case Level.ERROR =>
+          ctx.asScala.log.error(s"Supervisor $this saw failure:", unwrapped)
```
Wonder if this shouldn't include the exception like:
```
ctx.asScala.log.error(s"Supervisor $this saw failure: $unwrapped", unwrapped)
```

The Throwable parameter is used by logback to print the stack trace, but that is configurable (I believe).

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:
```diff
@@ -80,11 +72,16 @@ private abstract class AbstractSupervisor[I, Thr <: Throwable](strategy: Supervi
+        case Level.ERROR =>
+          ctx.asScala.log.error(s"Supervisor $this saw failure:", unwrapped)
+        case Level.WARN =>
+          ctx.asScala.log.warn(s"Supervisor $this saw failure:", unwrapped)
```
extract the `s"Supervisor $this saw failure:"` to a `val` since it's the same for all cases

patriknw commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala:
```diff
@@ -0,0 +1,474 @@
+ *
+ * If the `occurrences` is set to Int.MaxValue, no tracking is done.
+ */
+abstract class EventFilter(occurrences: Int) {
```
How far has this been tried? I see that it is used in a few tests with a SubstituteLogger. What is the idea for testing logging to any Logger (e.g. one given by ActorContext.log)?

Have we looked at any of existing testing utilities? Could it be an alternative to use something external instead of implementing this ourselves?
https://stackoverflow.com/questions/29076981/how-to-intercept-slf4j-logging-via-a-junit-test

franciscolopezsancho replied:
That's the approach I took. @raboof point me to that very same link, and it was quite useful; I have to say.

patriknw replied:
Ok, I just wanted to make sure that we invented the wheel again. I thought that we maybe could use http://projects.lidalia.org.uk/slf4j-test/ but there seems to be a few problems with that library:
* not maintained (last update 4 years ago)
* no license
* primarily intended for thread local logging (but also supports global log event recording)

Seems ok to continue on the approach taken here with the SubstituteLogger and our own testing api to assert the events. When we have confirmed that it works we should spend some time on the actual api of EventFilter, such as:
* javadsl/scaladsl
* simplify, remove unnecessary cruft

franciscolopezsancho replied:
:) yes. I tried to resist to implement ourselves but didn't found anything suitable.

Glad you agree on the approach. Let's do as you suggest. I'm looking forward to it.

patriknw reviewed (commented):
Skimmed this, and I'm curious to understand how far we have got here. How much is remaining? Any roadblocks?

franciscolopezsancho commented:
Hi @patriknw, I've been a bit blocked till today with testing MDC. The problem was adding it to the log message when configured through logback-test.xml (as @wsargent rightly suggested) vs. doing it directly through code (which was the way implemented before). No blocked anymore. I solved the rest of the comments except for a couple of them. I will review all I have, and push tomorrow or Friday what I've done. ASAP

Regards how much is left I think, and hope, that after these suggested changes the rest will be pretty smooth to change. After pushing what I have, I will check how many tests are failing, and we can make some measurement from there.

franciscolopezsancho commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorLoggingSpec.scala:
```diff
@@ -183,97 +172,41 @@ class ActorLoggingSpec extends ScalaTestWithActorTestKit("""
             context.log.debug("message")
             context.log.debug("{}", "arg1")
-            context.log.debug("{} {}", "arg1", "arg2")
+            context.log.debug("{} {}", 1, 2) //using Int to avoid ambiguous reference to overloaded definition
```
Had no idea 👍 Thanks!

franciscolopezsancho committed:
Solved suggestiong from https://github.com/akka/akka/pull/27324\#pullrequestreview-269703971 to https://github.com/akka/akka/pull/27324\#pullrequestreview-275042656 ,included

franciscolopezsancho committed:
Merge branch 'master' of https://github.com/akka/akka into typed-logging-with-slf4j

franciscolopezsancho commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala:
```diff
@@ -0,0 +1,474 @@
+object EventFilter {
+
+  /**
+   * Create a filter for Error events. Give up to one of <code>start</code> and <code>pattern</code>:
```
It seems to me — at least part of what it is supposed to do. I don't think that the description is complete though, as it can filter all level of messages. 

I added a mention to org.slf4j.event.LoggingEvent mention.

franciscolopezsancho commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala:
```diff
@@ -0,0 +1,474 @@
+   * Create a filter for Error events. Give up to one of <code>start</code> and <code>pattern</code>:
+   *
+   * {{{
+   * EventFilter[MyException]()                                         // filter only on exception type
```
I did the Queue part but not the examples `Logging.Error.NoCause.getClass`. 

I think it is a good idea, but I guess I'll try to pass first all the tests before I start to improve the current implementation. I still find it hard to grasp how all these testing classes work together.

franciscolopezsancho commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala:
```diff
@@ -289,12 +186,12 @@ private[akka] final class FunctionRef[-T](override val path: ActorPath, send: (T
    * Clear the log entries.
    */
-  def clearLog(): Unit = loggingAdapter.clearLog()
+  def clearLog(): Unit = ???
```
There's no equivalent in Slf4j. I'll keep it in mind. I still need to review tests that use this class.

franciscolopezsancho commented:
@wsargent @patriknw, I just pushed, and the result of the tests are:
```
[info] Total number of tests run: 768
[info] Suites: completed 87, aborted 0
[info] Tests: succeeded 681, failed 87, canceled 0, ignored 0, pending 15
```
I'll have a look at the failed tests to check out how easy it is to reuse what I've done in `LogMessagesSpec`. I'll keep you updated.

franciscolopezsancho committed:
added and appender that will intercept ch.qos.logback.classic.spi.ILoggingEvents, so still no completely slf4j generic. Started to pass tests in ActorLoggingSpec

franciscolopezsancho commented:
I added another means to intercept log messages as I didn't see any other way to catch them. This is extending `ch.qos.logback.core.AppenderBase`

patriknw commented:
I will continue this in https://github.com/akka/akka/pull/27552 since @franciscolopezsancho is on vacation.
